src/Provider/WFSProvider.js

```diff
@@ -89,7 +84,7 @@ function getFeatures(crs, tile, layer) {
                     throw err;
                 }
             })
-        .then(feature => assignLayer(layer.convert(feature), layer));
+        .then(feature => assignLayer(layer.convert(feature), layer)));
```

Comment:
This is a problem: you're putting into the global cache an entry that depends on a layer configuration.

You should either:
- not use the `url` as the key, but something that also depends on the layer
- or save 2 cache entries: `url -> json` (which can be shared) and then `key-that-depends-on-the-layer-id-and-the-json -> converted json`.

Reply:
I don't get what is wrong here. The `url` seems the most appropriate candidate here to know which tile we are storing. And isn't using 2 cache entries an overkill ? If I understand that would mean having:
```js
return Cache.get(urldAndID) || (Cache.get(urld) || Cache.set(urld, Fetcher.json()))
    .then(
        json => GeoJsonParser.parse(),
        err => throw new Error()
    ).then(feature => Cache.set(urldAndID, assignLayer(layer.convert(feature), layer))));
```

I feel that this is going back to the first version of this PR (in the part of `OGCWebServiceHelper`). WDYT ?

Reply:
> I don't get what is wrong here.

Let's say that I declare 2 layers using the same wfs datasource, but layer1.type = 'color' and layer2.type='geometry'.
With the current implementation, here's what happen: layer1 will fetch the features using one url and store in the cache the result of layer.convert (which is a texture). Later layer2 will fetch the features using the same url, so it will be a cache hit. Except that layer2 expect geometries and that cache.get(url) will return a texture.

Is it clearer?

Reply:
Oh yeah that makes sense, and the 2 cache entries even more. Trying it !

Reply:
As a related note, we should be quite careful about storing promises in cache containing `.then` with side-effects. Here, for example, if I'm not mistaken, user can customize `layer.convert`. We cannot guarantee that the user-provided function will be side-effect free. Maybe its result is time-dependant?

Maybe it'd be better to stick to caching request results for now (the most costly thing in term of user perceived time)?

Reply:
What do you mean by side-effects ? As long that it returns something compatible with the specified `update` function I see no problem, and the `update` function is specifiable by the user. I don't think we are taking a risk by storing promise here.

Reply:
We store promise with `.then` handlers, and these `then` callbacks will only be executed once with the cache mechanism, but in some cases they need to be executed each time.

Btw here, I just realized we already have a problem:

Let's say that I declare 2 layers using the same wfs datasource, but this time both are type 'geometry'.
With the current implementation, here's what happen: layer1 will fetch the features using one url and store in the cache the result of ` assignLayer(layer.convert(feature), layer))`, which is a Object3D with layer assigned to layer1. Later layer2 will fetch the features using the same url, so it will be a cache hit and layer2 will get a Object3D assigned to layer1.

It's because assignLayer has a side-effect. So imho: either we have no side-effect in the cache, or the key must depends on all the inputs, be explicit parameters or implicit context (meaning, all the side-effects of the attached `then` callbacks).

The latter is quite hard and fragile, because you need to adapt your key to potentially every changes of methods that are called in the `then` callback, supposing you got it right the first time (and the current thread shows it's not an easy task :-D), for each cases. That's why I think we should start by caching *only* the results of calls to `Fetcher` methods as a first step. So the pattern must be changed from:

```js
return Cache.get(...) || Cache.set(url, Fetcher.json()).then(...);
```
to
```js
return (Cache.get(...) || Cache.set(url, Fetcher.json(...))).then(...);
```

(notice than in the second case, `.then` operates on the result of `(Cache.get(...) || Cache.set(...))`)

Hope this is clearer :-)

Reply:
And getting the key right would be provider-dependant btw. I got actually beaten by that while trying this PR with the StaticProvider :-)

Reply:
With your suggestion, we could directly move the cache into the `Fetcher`, but the only thing we would gain is the lack of a new http request. Only caching the received file is not useful, as the browser already does that. But caching the processed resource, that is one thing that we can benefit from.

I feel that the struggle here is caching in the `WFSProvider`. I suggest that I revert back to the current `master` version, only replacing the `cache` by the new `Cache`, and do the definition of the key later. The initial goal of the PR is only to replace the `Map` we have in several classes, with a goal in mind to move the cache at a higher level later (maybe in the `Scheduler` ?).

Reply:
> but the only thing we would gain is the lack of a new http request

which is one of the main cause of latency, it's not a little gain.

> Only caching the received file is not useful, as the browser already does that

*If* the server is correctly configured...

I just mean: it's not completely useless :-)

But yeah, you're right: it would actually be a pity if we stopped at the request level for this cache... But at the very least caching should be done per-layer for WFS, because the results depends too much on the layer config.

No strong opinion on your last suggestion. I'd say: do what feels more comfortable to you :-)

Reply:
> If the server is correctly configured...

Yes indeed :sweat_smile: 

I'll let @peppsac or someone else give a last opinion on the subject, and I propose we do not take this PR too much further.